Refactor docker compose

ArturGajowy opened the pull request:
This PR aims at further deduplication of docker-compose files for product tests, following the path set by @brian-rickman in #5309. This will serve the following causes:
 - better readability and maintainability,
 - providing groundwork for running docker containers for product tests in EC2 by making volume definitions clearer.

This PR shifts from the [extending services](https://docs.docker.com/compose/extends/#extending-services) approach to the [multiple compose files](https://docs.docker.com/compose/extends/#multiple-compose-files) one. This allows for defining `volumes_from` and `depends_on` only once for each service (compared to multiple copies forced by the inheritance approach - see 'Note' in the 'extending services' link).

This can also be thought of as a switch from inheritance to composition. Each docker-compose.yml-file describes one aspect of the composition. Aspects applied later override those applied earlier. The application order - as can be seen in [`environment_docker_compose`](https://github.com/prestodb/presto/compare/master...Teradata:refactor-docker-compose?expand=1#diff-d5fa118c169436bb50bbd774dc84d440R95) function is as follows:

1. standard.yml
2. kerberos.yml (only for envs containing `kerberos` in name)
3. jdbc_db.yml
4. environment-specific .yml files

ArturGajowy commented:
@kokosing, @fiedukow please have a look :)

ArturGajowy commented:
Tested using 
```./run_on_docker.sh $e -x quarantine,big_query,profile_specific_tests```
for `$e` in
```
multinode
singlenode
singlenode-hdfs-impersonation
singlenode-kerberos-hdfs-impersonation
singlenode-kerberos-hdfs-no-impersonation
```

kokosing commented on presto-product-tests/conf/docker/common/kerberos.yml:
```diff
@@ -1,35 +1,16 @@
 version: '2'
```
do you still need this file in commons? how is it reused?

kokosing commented on presto-product-tests/bin/run_on_docker.sh:
```diff
@@ -70,25 +70,46 @@ function stop_application_runner_containers() {
+    -f "${ENVIRONMENT_DOCKER_COMPOSE_LOCATION}" \
+  )
+  docker-compose "${COMPOSE_FILES[@]}" "$@"
+}
```
maybe instead of this function, you could have a simple script in environment directory which would have specific for this environment docker-compose command. This give more copy-paste but
 - gives user possibility to run environment docker cluster without much thinking 
 - user can easy check the configuration (what overridden what) with - https://docs.docker.com/compose/reference/config/
 - it will be easier to run manual tests

kokosing replied:
Moreover this function may get overly complicated when we get more environments

arhimondr replied:
This is a bit hacked for the kerberos, but it is really nice for the regular environments. Standard hierarchy is declared here. Like `standard.yaml` -> `env_specific.yaml`. I would rather keep it as is.

ArturGajowy replied:
@kokosing and @fiedukow actually convinced me it's better outside of this file for usability reasons. Often we need to execute `docker-compose whatever` in a given directory / by passing a single `-f` argument and having this dependencies expressed in `run_on_docker` makes doing anything else than running the tests really hard.

I'm hesitating between two approaches: having a `compose.sh` oneliner in each of env dirs or writing a `docker-composer` script in `bin` that would read `#requires ../common/standard.yml` directives from the files recursively. I think I'll start with the former.

kokosing commented:
Update documentation, especially part with manual tests execution. Make sure you are still able to debug presto or product tests easily during tests.

arhimondr commented on presto-product-tests/conf/docker/common/standard.yml:
```diff
@@ -8,16 +8,10 @@ services:
     extends:
-      service: hadoop-base
+      service: java-8-base
+    image: 'teradatalabs/cdh5-hive:2'
```
Why do we need to change the image here to `cdh5-hive`? In non-kerberized environment Presto must be run within the simple `java8` container.

arhimondr commented on presto-product-tests/conf/docker/common/standard.yml:
```diff
@@ -27,22 +21,23 @@ services:
+    depends_on:
+      - presto-master
+    volumes_from:
+      - presto-master
```
new line

arhimondr commented on presto-product-tests/conf/docker/singlenode-hdfs-impersonation/docker-compose.yml:
```diff
@@ -1,35 +1,7 @@
-    extends:
-      file: ../common/standard.yml
-      service: presto-master-common
     command: /docker/volumes/presto/presto-product-tests/conf/docker/files/presto-launcher-wrapper.sh singlenode run
```
I would move it to `standard.yaml`. It is different in kerberos/multinode and it is already overwritten there.

arhimondr commented on presto-product-tests/conf/docker/common/standard.yml:
```diff
@@ -27,22 +21,23 @@ services:
       service: java-8-base
     hostname: presto-master
     volumes:
       - ../../../../presto-product-tests/conf/tempto/tempto-configuration-for-docker-default.yaml:/docker/volumes/tempto/tempto-configuration-local.yaml
```
why do we need `tempto-configuration` on `presto-master`? Can't we keep it on `application-runner`?

ArturGajowy replied:
I think yes, I'll try to move it

arhimondr commented on presto-product-tests/conf/docker/common/standard.yml:
```diff
@@ -27,22 +21,23 @@ services:
     extends:
       service: java-8-base
     hostname: application-runner
+    depends_on:
```
It shouldn't depend on `presto-master`. Application runner is independent container to run the ad-hoc application, and shouldn't be run as a deamon.

arhimondr commented on presto-product-tests/conf/docker/common/kerberos.yml:
```diff
@@ -1,35 +1,16 @@
+    image: 'teradatalabs/cdh5-hive-kerberized:3'
     command: /docker/volumes/presto/presto-product-tests/conf/docker/files/presto-launcher-wrapper.sh singlenode-kerberized run
     volumes:
       - ../../../../presto-product-tests/conf/tempto/tempto-configuration-for-docker-kerberos.yaml:/docker/volumes/tempto/tempto-configuration-local.yaml
```
same question about the tempto configuration

arhimondr commented on presto-product-tests/conf/docker/singlenode/docker-compose.yml:
```diff
@@ -1,33 +1,5 @@
-    extends:
-      file: ../common/standard.yml
-      service: presto-master-common
     command: /docker/volumes/presto/presto-product-tests/conf/docker/files/presto-launcher-wrapper.sh singlenode run
```
remove it once you move it to the `standard`

arhimondr commented on presto-product-tests/bin/run_on_docker.sh:
```diff
@@ -70,25 +70,46 @@ function stop_application_runner_containers() {
 
+function stop_all_containers() {
+  local ENVIRONMENT
+  for ENVIRONMENT in $(getAvailableEnvironments)
```
Maybe i'm wrong, but as i know `local` is visible only within the function where it is declared. If i'm right the `environment_docker_compose` in `stop_docker_compose_containers` will use global `ENVIRONMENT` variable and will stop the user selected environment for the 4 times. I'm not a big bash specialist, so i could be wrong. But you better check it.

ArturGajowy replied:
Checked it :) `local` means roughly 'make changes to this variable not leak to outer scopes; inner scopes are going to see them'.

arhimondr commented on presto-product-tests/bin/run_on_docker.sh:
```diff
@@ -70,25 +70,46 @@ function stop_application_runner_containers() {
+  done
+}
+
 function stop_docker_compose_containers() {
```
I think it is better to pass explicitly the environment you want to stop the containers for.

arhimondr commented:
Generally looks good. But we need to make sure that our user docs are up to date, and the product tests are passing on:

1. Mac/Linux
2. run_on_docker.sh (all environments)
3. run.sh with spinning up the Presto and Hadoop following the our manual.

ArturGajowy commented:
@arhimondr @fiedukow @kokosing Please have another look. I addressed your comments and rebased @ master. Running product tests for all environments now.

electrum commented:
Some of the commit titles are too long. Please limit to ~50 characters (no more than 60). Note how GitHub truncates them.

ArturGajowy commented:
@electrum fixed :)

kokosing commented on presto-product-tests/bin/run_on_docker.sh:
```diff
@@ -70,25 +71,39 @@ function stop_application_runner_containers() {
+
+function environment_docker_compose() {
+  local ENVIRONMENT_COMPOSE_SCRIPT_LOCATION="${DOCKER_CONF_LOCATION}/${ENVIRONMENT}/compose.sh"
+  ${ENVIRONMENT_COMPOSE_SCRIPT_LOCATION} "$@"
```
inline this function to global variable

kokosing commented on presto-product-tests/bin/run_on_docker.sh:
```diff
@@ -115,25 +130,24 @@ function terminate() {
 }
 
 ENVIRONMENT=$1
 SCRIPT_DIR=$(dirname $(absolutepath "$0"))
```
maybe you can replace this with `BASH_SOURCE` usage?

kokosing commented on presto-product-tests/conf/docker/multinode/compose.sh:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+docker-compose \
+-f ${BASH_SOURCE%/*}/../common/standard.yml \
```
nice trick ;)

kokosing commented:
lgtm

ArturGajowy commented:
@electrum back to you :)

electrum commented:
This looks good, but there were merge conflicts. Can you rebase and assign back to me?

ArturGajowy commented:
Rebased + ran product tests in all profiles (they pass ;))

ArturGajowy commented:
@electrum it's ready for merging :)

electrum commented:
Merged, thanks!
